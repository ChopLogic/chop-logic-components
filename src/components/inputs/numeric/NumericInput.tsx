import React from 'react';
import { useElementIds } from 'hooks/use-element-ids';

import ChopLogicErrorMessage from 'components/inputs/_common/error-message/ErrorMessage';
import ChopLogicLabel from 'components/inputs/_common/label/Label';
import { CLIcon } from 'components/misc/icon/Icon';

import InputInnerButton from '../_common/input-inner-button/InputInnerButton';

import { useChopLogicNumericInputController } from './helpers';
import {
  StyledButtonsWrapper,
  StyledFieldWrapper,
  StyledNumericInput,
  StyledNumericInputContainer,
  StyledNumericInputWrapper,
} from './NumericInput.styled';

export type NumericValidationFunction = (input?: number) => boolean;

export type ChopLogicNumericInputProps = React.InputHTMLAttributes<HTMLInputElement> & {
  name: string;
  label: string;
  errorMessage?: string;
  validator?: NumericValidationFunction;
  clearable?: boolean;
  onClear?: () => void;
  onSpinButtonClick?: (value?: number) => void;
};

const ChopLogicNumericInput: React.FC<ChopLogicNumericInputProps> = ({
  name,
  label,
  errorMessage,
  onChange,
  onClear,
  validator,
  min,
  max,
  disabled = false,
  required = false,
  readOnly = false,
  clearable = false,
  defaultValue = 0,
  step = 1,
  ...props
}) => {
  const { elementId, errorId } = useElementIds(props?.id);
  const { value, valid, handleChange, handleClear, increment, decrement, minValue, maxValue } = useChopLogicNumericInputController({
    name,
    defaultValue,
    min,
    max,
    onChange,
    onClear,
    required,
    validator,
    step: +step,
  });

  return (
    <StyledNumericInputContainer className={props?.className} style={props?.style}>
      <StyledNumericInputWrapper $disabled={disabled} $invalid={!valid}>
        <ChopLogicLabel label={label} required={required} inputId={elementId} />
        <StyledFieldWrapper>
          <StyledNumericInput
            id={elementId}
            name={name}
            type='number'
            disabled={disabled}
            required={required}
            readOnly={readOnly}
            placeholder='0'
            aria-invalid={!valid}
            aria-errormessage={errorId}
            value={value.toString()}
            onChange={handleChange}
            min={minValue}
            max={maxValue}
            step={step}
            onBlur={props?.onBlur}
            onFocus={props?.onFocus}
            tabIndex={props?.tabIndex}
          />
          <StyledButtonsWrapper>
            {clearable && <InputInnerButton onClick={handleClear} label={`Clear input for ${label}`} icon={CLIcon.Remove} />}
            <InputInnerButton onClick={increment} label={`Increment value for ${label}`} icon={CLIcon.Up} />
            <InputInnerButton onClick={decrement} label={`Decrement value for ${label}`} icon={CLIcon.Down} />
          </StyledButtonsWrapper>
        </StyledFieldWrapper>
      </StyledNumericInputWrapper>
      <ChopLogicErrorMessage errorId={errorId} message={errorMessage} visible={!valid} />
    </StyledNumericInputContainer>
  );
};

export default ChopLogicNumericInput;
