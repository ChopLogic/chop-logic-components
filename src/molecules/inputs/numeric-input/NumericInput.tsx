import { Label } from '@atoms';
import Button from '@atoms/button/Button.tsx';
import { ButtonView, IconName } from '@enums';
import { useElementIds } from '@hooks';
import { NumericInputProps } from '@models';
import { getClassName } from '@utils';
import { FC } from 'react';

import { ErrorMessage, Input } from '../__elements__/index.ts';
import { useNumericInputController } from './NumericInput.controller.ts';
import styles from './NumericInput.module.scss';

const NumericInput: FC<NumericInputProps> = ({
  name,
  label,
  errorMessage,
  onChange,
  onBlur,
  onFocus,
  onSpinButtonClick,
  validator,
  min,
  max,
  id,
  tabIndex,
  disabled = false,
  required = false,
  readOnly = false,
  hasSpinButtons = true,
  defaultValue = 0,
  step = 1,
  className,
  ...rest
}) => {
  const { elementId, errorId } = useElementIds(id);
  const { value, valid, handleChange, increment, decrement, minValue, maxValue } = useNumericInputController({
    name,
    defaultValue,
    min,
    max,
    onChange,
    required,
    validator,
    step,
    onSpinButtonClick,
  });
  const inputClass = getClassName([styles.numeric, className]);

  return (
    <div {...rest} className={inputClass}>
      <Label label={label} required={required} inputId={elementId} />
      <Input
        id={elementId}
        name={name}
        type='number'
        disabled={disabled}
        required={required}
        readOnly={readOnly}
        placeholder='0'
        aria-invalid={!valid}
        aria-errormessage={errorId}
        value={value.toString()}
        onChange={handleChange}
        min={minValue}
        max={maxValue}
        step={step}
        onBlur={onBlur}
        onFocus={onFocus}
        tabIndex={tabIndex}
      >
        {hasSpinButtons && (
          <span>
            <Button
              onClick={decrement}
              view={ButtonView.Inner}
              label={`Decrement value for ${label}`}
              icon={IconName.ChevronLeft}
              disabled={disabled}
            />
            <Button
              onClick={increment}
              view={ButtonView.Inner}
              label={`Increment value for ${label}`}
              icon={IconName.ChevronRight}
              disabled={disabled}
            />
          </span>
        )}
        <ErrorMessage errorId={errorId} message={errorMessage} visible={!valid} className={styles.errorMessage} />
      </Input>
    </div>
  );
};

export default NumericInput;
