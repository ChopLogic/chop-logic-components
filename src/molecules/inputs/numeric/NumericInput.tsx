import { ChopLogicButtonView, ChopLogicIconName } from '@enums';
import { useElementIds } from '@hooks';
import { ChopLogicNumericInputProps } from '@models';
import ChopLogicButton from '@molecules/inputs/button/Button.tsx';
import { getClassName } from '@utils';
import { FC } from 'react';

import { ErrorMessage, Input, Label } from '../_elements';
import { useChopLogicNumericInputController } from './NumericInput.controller.ts';
import styles from './NumericInput.module.scss';

const ChopLogicNumericInput: FC<ChopLogicNumericInputProps> = ({
  name,
  label,
  errorMessage,
  onChange,
  onBlur,
  onFocus,
  onSpinButtonClick,
  validator,
  min,
  max,
  id,
  tabIndex,
  disabled = false,
  required = false,
  readOnly = false,
  hasSpinButtons = true,
  defaultValue = 0,
  step = 1,
  className,
  ...rest
}) => {
  const { elementId, errorId } = useElementIds(id);
  const { value, valid, handleChange, increment, decrement, minValue, maxValue } = useChopLogicNumericInputController({
    name,
    defaultValue,
    min,
    max,
    onChange,
    required,
    validator,
    step,
    onSpinButtonClick,
  });
  const inputClass = getClassName([styles.numeric, className]);

  return (
    <div {...rest} className={inputClass}>
      <Label label={label} required={required} inputId={elementId} />
      <Input
        id={elementId}
        name={name}
        type='number'
        disabled={disabled}
        required={required}
        readOnly={readOnly}
        placeholder='0'
        aria-invalid={!valid}
        aria-errormessage={errorId}
        value={value.toString()}
        onChange={handleChange}
        min={minValue}
        max={maxValue}
        step={step}
        onBlur={onBlur}
        onFocus={onFocus}
        tabIndex={tabIndex}
      >
        {hasSpinButtons && (
          <span>
            <ChopLogicButton
              onClick={decrement}
              view={ChopLogicButtonView.Inner}
              label={`Decrement value for ${label}`}
              icon={ChopLogicIconName.ChevronLeft}
              disabled={disabled}
            />
            <ChopLogicButton
              onClick={increment}
              view={ChopLogicButtonView.Inner}
              label={`Increment value for ${label}`}
              icon={ChopLogicIconName.ChevronRight}
              disabled={disabled}
            />
          </span>
        )}
        <ErrorMessage errorId={errorId} message={errorMessage} visible={!valid} className={styles.errorMessage} />
      </Input>
    </div>
  );
};

export default ChopLogicNumericInput;
