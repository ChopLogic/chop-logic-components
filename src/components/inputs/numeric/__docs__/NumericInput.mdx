import { Canvas, Meta } from '@storybook/blocks';
import * as NumericInput from './NumericInput.stories';

<Meta of={NumericInput} title='NumericInput' />

# ChopLogicNumericInput

Numeric input component with standard HTML attributes.

### Example

<Canvas of={NumericInput.Default} />

### Usage

```tsx
import { ChopLogicNumericInput } from 'chop-logic-components';

const Example = () => {
  return (
    <ChopLogicNumericInput
      label='Enter your age'
      name='age'
      errorMessage='Enter a number between 1 and 99'
      validator={(age) => !!age && age >= 1 && age < 100}
    />
  );
};

export default Example;
```

### Props

| Prop                  | Type                       | Description                                                                                                                              | Default     | Required |
|-----------------------|----------------------------|------------------------------------------------------------------------------------------------------------------------------------------|-------------|----------|
| **name**              | `string`                   | A string representing the name attribute of the component, used for form submissions.                                                    | `undefined` | Yes      |
| **label**             | `string`                   | A string that represents the label for the component.                                                                                    | `undefined` | Yes      |
| **validator**         | `(value?: number) => void` | A function or regular expression that allows you to validate an input value and display an error message if validation fails.            | `undefined` | No       |
| **errorMessage**      | `string`                   | An optional string that displays when the input value is incorrect.                                                                      | `undefined` | No       |
| **onChange**          | `ChangeEventHandler`       | A function that is called when a user type something in the input. It receives a ChangeEventHandler for handling the event on the input. | `undefined` | No       |
| **onFocus**           | `FocusEventHandler`        | A function that is called when focus is set on the input.                                                                                | `undefined` | No       |
| **onBlur**            | `FocusEventHandler`        | A function that is called when focus leaves the input.                                                                                   | `undefined` | No       |
| **onSpinButtonClick** | `(value?: number) => void` | A function that is called when one of the spin buttons is clicked.                                                                       | `undefined` | No       |
| **step**              | `number`                   | The number by which the input value is incremented/decremented when pressing keyboard keys.                                              | `1`         | No       |
| **max**               | `number`                   | Maximum number in the input.                                                                                                             | `undefined` | No       |
| **min**               | `number`                   | Minimum number in the input.                                                                                                             | `undefined` | No       |
| **defaultValue**      | `number`                   | Initial number in the input.                                                                                                             | `0`         | No       |
| **hasSpinButtons**    | `boolean`                  | A boolean indicating whether to display the spin buttons.                                                                                | `true`      | No       |
| **disabled**          | `boolean`                  | When disabled, the input cannot be interacted with.                                                                                      | `false`     | No       |
| **required**          | `boolean`                  | When required, the input must be filled before form submission.                                                                          | `false`     | No       |
| **readOnly**          | `boolean`                  | When readonly, the input cannot be changed.                                                                                              | `false`     | No       |
| **id**                | `string`                   | The id global attribute defines an identifier (ID) which must be unique in the whole document.                                           | `undefined` | No       |
| **className**         | `string`                   | The class global attribute is a list of the classes of the element.                                                                      | `undefined` | No       |
| **style**             | `CSSProperties`            | The style global attribute contains CSS styling declarations to be applied to the element.                                               | `undefined` | No       |
| **tabIndex**          | `number`                   | The tabindex global attribute allows developers to make HTML elements focusable.                                                         | `undefined` | No       |
| **title**             | `string`                   | The title global attribute contains text representing advisory information related to the element it belongs to.                         | `undefined` | No       |

### Types

```ts
export type ChopLogicNumericInputProps = CommonInputProps & {
  errorMessage?: string;
  validator?: NumericValidationFunction;
  hasSpinButtons?: boolean;
  min?: number;
  max?: number;
  step?: number;
  readOnly?: boolean;
  defaultValue?: number;
  onSpinButtonClick?: (value?: number) => void;
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  onBlur?: React.FocusEventHandler<HTMLInputElement>;
  onFocus?: React.FocusEventHandler<HTMLInputElement>;
};

export type NumericValidationFunction = (input?: number) => boolean;

export type CommonInputProps = CommonComponentProps & {
  label: string;
  name: string;
  disabled?: boolean;
  required?: boolean;
};

export type CommonComponentProps = {
  id?: string;
  className?: string;
  style?: React.CSSProperties;
  tabIndex?: number;
  title?: string;
  theme?: 'dark' | 'light';
};
```
