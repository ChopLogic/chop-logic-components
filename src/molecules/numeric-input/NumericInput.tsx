import { ErrorMessage, Input, Label } from '@atoms';
import Button from '@atoms/button/Button';
import { ButtonView, IconName } from '@enums';
import { useElementIds } from '@hooks';
import type { NumericInputProps } from '@models';
import { getClassName } from '@utils';
import type { FC } from 'react';

import { useNumericInputController } from './NumericInput.controller';
import styles from './NumericInput.module.scss';

const NumericInput: FC<NumericInputProps> = ({
  name,
  label,
  errorMessage,
  onChange,
  onBlur,
  onFocus,
  onSpinButtonClick,
  validator,
  min,
  max,
  id,
  tabIndex,
  disabled = false,
  required = false,
  readOnly = false,
  hasSpinButtons = true,
  defaultValue = 0,
  step = 1,
  className,
  ...rest
}) => {
  const { elementId, errorId } = useElementIds(id);
  const { value, valid, handleChange, increment, decrement, minValue, maxValue } =
    useNumericInputController({
      name,
      defaultValue,
      min,
      max,
      onChange,
      required,
      validator,
      step,
      onSpinButtonClick,
    });
  const inputClass = getClassName([styles.numeric, className]);

  return (
    <div {...rest} className={inputClass}>
      <Label label={label} required={required} inputId={elementId} />
      <Input
        id={elementId}
        name={name}
        type="number"
        disabled={disabled}
        required={required}
        readOnly={readOnly}
        placeholder="0"
        aria-invalid={!valid}
        aria-errormessage={errorId}
        value={value.toString()}
        onChange={handleChange}
        min={minValue}
        max={maxValue}
        step={step}
        onBlur={onBlur}
        onFocus={onFocus}
        tabIndex={tabIndex}
      >
        {hasSpinButtons && (
          <span>
            <Button
              onClick={decrement}
              view={ButtonView.Inner}
              label={`Decrement value for ${label}`}
              icon={IconName.ChevronLeft}
              disabled={disabled}
            />
            <Button
              onClick={increment}
              view={ButtonView.Inner}
              label={`Increment value for ${label}`}
              icon={IconName.ChevronRight}
              disabled={disabled}
            />
          </span>
        )}
        <ErrorMessage
          errorId={errorId}
          message={errorMessage}
          visible={!valid}
          className={styles.errorMessage}
        />
      </Input>
    </div>
  );
};

export default NumericInput;
