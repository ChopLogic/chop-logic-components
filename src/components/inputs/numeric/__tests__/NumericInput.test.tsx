import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import ChopLogicForm from 'components/containers/form/Form';

import NumericInput from '../NumericInput';

describe('NumericInput', () => {
  const testProps = {
    id: 'numeric-input',
    name: 'quantity',
    label: 'Quantity',
    style: { width: '500px' },
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(<NumericInput {...testProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with default props', () => {
    render(<NumericInput {...testProps} />);

    // Check if the label is rendered
    expect(screen.getByText(/Quantity/i)).toBeInTheDocument();

    // Check if the input is rendered
    const input = screen.getByRole('spinbutton');
    expect(input).toBeInTheDocument();
    expect(input).toHaveAttribute('type', 'number');
  });

  it('displays the correct default value', () => {
    render(<NumericInput {...testProps} defaultValue={5} />);

    const input = screen.getByRole('spinbutton');
    expect(input).toHaveValue(5);
  });

  it('updates value when typing', () => {
    const handleChange = vi.fn();
    render(<NumericInput {...testProps} onChange={handleChange} />);

    const input = screen.getByRole('spinbutton');
    fireEvent.change(input, { target: { value: '10' } });

    expect(input).toHaveValue(10);
    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  it('applies disabled state correctly', () => {
    render(<NumericInput {...testProps} disabled />);

    const input = screen.getByRole('spinbutton');
    expect(input).toBeDisabled();
  });

  it('displays error message', () => {
    render(<NumericInput {...testProps} errorMessage='Invalid input' />);

    expect(screen.getByText(/Invalid input/i)).toBeInTheDocument();
  });

  it('shows required label', () => {
    render(<NumericInput {...testProps} required />);

    expect(screen.getByText('*')).toBeInTheDocument();
    expect(screen.getByRole('spinbutton')).toBeRequired();
  });

  it('handles min and max constraints', () => {
    render(<NumericInput {...testProps} min={1} max={10} />);

    const input = screen.getByRole('spinbutton');
    fireEvent.change(input, { target: { value: '-1' } });
    expect(input).toHaveValue(-1);
    expect(input).toBeInvalid();

    fireEvent.change(input, { target: { value: '11' } });
    expect(input).toHaveValue(11);
    expect(input).toBeInvalid();
  });

  it('does not allow non-numeric input', () => {
    render(<NumericInput {...testProps} />);

    const input = screen.getByRole('spinbutton');
    fireEvent.change(input, { target: { value: 'abc' } });

    expect(input).toHaveValue(0); // If non-numeric input is entered, it should revert to 0.
  });

  it('should respect step increment', () => {
    render(<NumericInput {...testProps} step={2} defaultValue={0} />);

    const input = screen.getByRole('spinbutton');
    fireEvent.change(input, { target: { value: '4' } });

    expect(input).toHaveValue(4);
  });

  it('should allow step increment using custom spin buttons', async () => {
    render(<NumericInput {...testProps} step={3} defaultValue={0} />);

    await userEvent.click(screen.getByLabelText(`Increment value for ${testProps.label}`));
    await userEvent.click(screen.getByLabelText(`Increment value for ${testProps.label}`));

    expect(screen.getByRole('spinbutton')).toHaveValue(6);
  });

  it('should allow step decrement using custom spin buttons', async () => {
    render(<NumericInput {...testProps} step={5} defaultValue={0} />);

    await userEvent.click(screen.getByLabelText(`Decrement value for ${testProps.label}`));
    await userEvent.click(screen.getByLabelText(`Decrement value for ${testProps.label}`));

    expect(screen.getByRole('spinbutton')).toHaveValue(-10);
  });

  it('should call onSpinButtonClick', async () => {
    const mockedCallback = vi.fn();
    render(<NumericInput {...testProps} onSpinButtonClick={mockedCallback} />);

    await userEvent.click(screen.getByLabelText(`Increment value for ${testProps.label}`));
    await userEvent.click(screen.getByLabelText(`Decrement value for ${testProps.label}`));

    expect(mockedCallback).toHaveBeenCalledTimes(2);
  });

  it('should take an initial value from the form context', async () => {
    render(
      <ChopLogicForm initialValues={{ quantity: 13 }}>
        <NumericInput {...testProps} />
      </ChopLogicForm>,
    );
    const input = screen.getByRole('spinbutton');
    expect(input).toHaveValue(13);
  });
});
