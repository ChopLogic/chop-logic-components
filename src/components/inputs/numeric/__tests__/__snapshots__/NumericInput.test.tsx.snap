// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NumericInput > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-knesRu keCzej"
  >
    <div
      class="sc-dnaUSb hVscX"
    >
      <label
        class="sc-iKTcqh dCKerg"
        for="numeric-input"
      >
        <span
          class="sc-gnpbhQ eBqzwP"
        >
          Quantity
        </span>
      </label>
      <span
        class="sc-dkjaqt lfyCkO"
      >
        <input
          aria-errormessage="numeric-input_error"
          aria-invalid="false"
          class="sc-kkmypM fcxFcU"
          id="numeric-input"
          max="9007199254740991"
          min="-9007199254740991"
          name="quantity"
          placeholder="0"
          step="1"
          type="number"
          value="0"
        />
        <span
          class="sc-jCbFiK fEirZx"
        >
          <button
            aria-label="Increment value for Quantity"
            class="sc-iCZwEW dTVEfm"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              viewBox="0 0 16 16"
              width="1em"
            >
              <path
                d="M7.776 5.553a.5.5 0 01.448 0l6 3a.5.5 0 11-.448.894L8 6.56 2.224 9.447a.5.5 0 11-.448-.894l6-3z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <button
            aria-label="Decrement value for Quantity"
            class="sc-iCZwEW dTVEfm"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              viewBox="0 0 16 16"
              width="1em"
            >
              <path
                d="M1.553 6.776a.5.5 0 01.67-.223L8 9.44l5.776-2.888a.5.5 0 11.448.894l-6 3a.5.5 0 01-.448 0l-6-3a.5.5 0 01-.223-.67z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </span>
      </span>
    </div>
    <span
      class="sc-dsLQwm Oiwth"
      id="numeric-input_error"
    >
      Invalid input
    </span>
  </div>
</DocumentFragment>
`;
